design/ceph/object/ceph-bucket-notification-crd.md

```diff
@@ -0,0 +1,82 @@
+# Ceph bucket notifications CRD
+
+## Overview
+
+Ceph added support for the bucket notifications feature from Nautilus onwards. It allows to send messages to various endpoints when a new event occurs on a bucket (ref: <https://docs.ceph.com/docs/master/radosgw/notifications/)>
+
+Setup of those notifications are normally done by sending HTTP requests to the RGW, either to create/delete topics pointing to specific endpoints, or create/delete bucket notifications based on those topics.
+
+This functionality eases this process by avoiding to use external tools or scripts. It is replaced by creation of CR definitions that contain all the information necessary to create topics and/or notifications, which the rook operator processes.
+
+## Goals
+
+Creates a CRD for topics and a CRD for notifications, defining all the necessary and optional information for the various endpoints.
+
+Extends the rook operator to handle the CRs that would be submitted by users.
+
+## Implementation
+
+The CR for a topic configuration takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephBucketTopic
+metadata:
+  name: # name of the topic
+  namespace: # namespace where topic belongs
+Spec:
+  endpoint: #(mandatory) URI of an endpoint to send push notification to
+  opaqueData: #(optional) opaque data is set in the topic configuration
```

Comment:
In what format is the opaqueData?

Reply:
any string

Reply:
To Travis previous point, examples would be good. For example, depending on the framework or solution, I never know if `endoint` will require http in front or not...

Reply:
@guimou from this CRD perspective we don't know what lies beyond the endpoint. If the endpoint is defined to be HTTP, it may be just an adapter to another system (e.g. knative). However, we are agnostic to that here.

Setting up an end2end solution would be outside the scope of this PR (and rook in general).
E.g. and operator that:
- creates topic and bucket notification CRs
- creates knative ceph source CR matching the above topic
- setup knative serving to consume the notifications

Reply:
> In what format is the opaqueData?

little more info on that...
- there could be cases where ceph is used by several applications, and the message bus (e.g. kafka) used to route the notifications is also shared by these applications. usually different kafka topics would be used for that, but if topics are also shared, the applications may find it useful to add their identity in that field, so that the consumer would know which app requested for the notification.
- another case is where the same message bus serves multiple ceph clusters, and opaque-data could hold the identity (or entry point) of the cluster that originated the notification, so that if the consumer needs to fetch the object it was notified on, it would know where to get it from